src/Condition.js

```diff
@@ -0,0 +1,240 @@
+
+/**
+ * Determines how to transform children configuration based on the affix type.
+ */
+const configMapper = {
+  prefix: cfg => transform(cfg),
+  infix: cfg => cfg.map(child => transform(child)),
+};
+
+/**
+ * Determines how to compose VCL based on the affix type.
+ */
+const vclComposer = {
+  prefix: (item, op) => `${op}(${item.toVCL()})`,
+  infix: (items, op) => `(${items.map(item => item.toVCL()).join(op)})`,
+};
+
+/**
+ * Boolean conditions
+ */
+const booleanMap = {
+  or: {
+    mapper: configMapper.infix,
+    op: ' || ',
+    vcl: vclComposer.infix,
+    express: (items, req) => items.reduce((result, item) => result || item.evaluate(req), false),
+  },
+  and: {
+    mapper: configMapper.infix,
+    op: ' && ',
+    vcl: vclComposer.infix,
+    express: (items, req) => items.reduce((result, item) => result && item.evaluate(req), true),
+  },
+  not: {
+    mapper: configMapper.prefix,
+    op: ' !',
+    vcl: vclComposer.prefix,
+    express: (item, req) => !item.evaluate(req),
+  },
+};
+
+class BooleanCondition {
+  constructor(entry, cfg) {
+    this._entry = entry;
+    this._items = this._entry.mapper(cfg);
+  }
+
+  toVCL() {
+    return this._entry.vcl(this._items, this._entry.op);
+  }
+
+  evaluate(req) {
+    return this._entry.express(this._items, req);
+  }
+}
+
+/**
+ * PropertyCondition
+ */
+class PropertyCondition {
+  constructor(prop, op, value, name) {
+    if (op && prop.allowed_ops.indexOf(op) === -1) {
+      throw new Error(`Property ${name} does not support operation: ${op}`);
+    }
+
+    this._prop = prop;
+    this._op = op;
+    this._value = value;
+    this._name = name;
+  }
+
+  toVCL() {
+    const quote = this._prop.type === 'string' ? '"' : '';
+    let value = this._value;
+    let op = this._op;
+
+    if (quote === '"' && !op) {
+      // substring-start
+      value = `^${value}`;
+      op = '~';
+    }
+    if (!op || op === '=') {
+      // operand defaults to equal
+      op = '==';
+    }
+    const vcl = typeof this._prop.vcl === 'function' ? this._prop.vcl(this) : this._prop.vcl;
+    return `${vcl} ${op} ${quote}${value}${quote}`;
+  }
+
+  evaluate(req) {
+    const actual = this._prop.express(req, this);
+    const value = this._value;
+
+    if (!actual) {
+      return false;
+    }
+
+    switch (this._op) {
+      case '=':
+        return actual === value;
+      case '~':
+        return !!actual.match(value);
+      case '<':
+        return actual < value;
+      case '>':
+        return actual > value;
+      default:
+        return actual.startsWith(value);
+    }
+  }
+
+  get name() {
+    return this._name;
+  }
+
+  get type() {
+    return this._prop.type;
+  }
+}
+
+/**
+ * Known properties
+ */
+const propertyMap = {
+  url: {
+    vcl: 'req.http.X-Full-URL',
+    express: req => `${req.protocol}://${req.hostname}/${req.path}`,
+    type: 'string',
+    allowed_ops: '=~',
+  },
+  'url.hostname': {
+    vcl: 'req.http.host',
+    express: req => req.hostname,
```

Comment:
I think the express [req.hostname](https://expressjs.com/en/4x/api.html#req.hostname) is wrong here, as it doesn't include the port. I would use `req.headers.host` instead.